src/Week5/Calc.hs

```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE FlexibleInstances    #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+
+
+module Week5.Calc where
+
+import Week5.ExprT (ExprT (..))
+import Week5.Parser (parseExp)
+import qualified Week5.StackVM as VM (Program, StackExp (..), stackVM)
+
+import Control.Applicative (liftA2)
+
+---------- ex 1 ----------
+eval :: ExprT -> Integer
+eval (Lit n)   = n
+eval (Add x y) = eval x + eval y
+eval (Mul x y) = eval x * eval y
+
+---------- ex 2 ----------
+evalStr :: String -> Maybe Integer
+evalStr str = eval <$> parseExp Lit Add Mul str
+
+---------- ex 3 ---------
+class Expr a where
+    lit :: Integer -> a
+    add :: a -> a -> a
+    mul :: a -> a -> a
+
+instance Expr ExprT where
+    lit = Lit
+    add = Add
+    mul = Mul
+
+---------- ex 4 ----------
+instance Expr Integer where
+    lit = id
+    add = (+)
+    mul = (*)
+
+instance Expr Bool where
+    lit = (<0)
+    add = (||)
+    mul = (&&)
+
+newtype MinMax = MinMax Integer
+    deriving (Eq, Show)
+
+instance Expr MinMax where
+    lit = MinMax
+    add (MinMax x) (MinMax y) = MinMax $ max x y
+    mul (MinMax x) (MinMax y) = MinMax $ min x y
+
+newtype Mod7 = Mod7 Integer
+    deriving (Eq, Show)
+
+instance Expr Mod7 where
+    lit = Mod7 . (`mod` 7)
+    add (Mod7 x) (Mod7 y) = lit $ x + y
+    mul (Mod7 x) (Mod7 y) = lit $ x * y
+
+-- For types like MinMax and Mod7 I think I should be able
+-- to use something similar to "lift" to avoid patternmatching
+-- I think liftA2 has the right type signature
+-- liftA2 :: (a -> b -> c) -> f a -> f b -> f c
+-- but MinMax is not recognised as a functor (I think?)
+-- Ideally I'd want to write something like this:
+--     add x y = liftA2 (max) x y
+-- how do I do this?
```

Comment:
This is a tricky question. Unfortunately, you can't use `liftA2` with types like `Mod7` and `MinMax`. Arguments of `liftA2` have type `f a`. This requires for the types to be higher kinded (it supposed to be the next lecture). Or, in another words, it works only for parametrized types. If instead of:

```haskell
newtype MinMax = MinMax Integer
```

you had

```haskell
newtype MinMax a = MinMax a
```

you could write more general instance:

```haskell
instance Ord a => Expr (MinMax a) where
```

And after you implement `Functor` and `Applicative` instances for `MinMax` you should be able to write:

```
add = liftA2 max  -- note, you can eta-reduce here
```

But `Functor` and `Applicative` are later topics. Don't need to dive into more complicated type hierarchy too soon :slightly_smiling_face: